GVRf/Framework/src/org/gearvrf/GVRHybridObject.java

```diff
@@ -15,209 +15,97 @@
-     * for every instance of every object that descends from
-     * {@link GVRSceneObject}! Most wrappers are kept alive by references in the
-     * scene graph; when you remove them from the scene graph, they can
-     * generally be dereferenced and (eventually) garbage collected without any
-     * issues.
+     * Most objects will be automatically closed when Java's garbage collector
+     * detects that they are no longer being used: Explicitly closing an object
+     * that's still linked into the scene graph will almost certainly crash your
+     * GVRF app. You should only {@code close()} transient objects (especially
+     * those that use lots of memory, like large textures) that you
+     * <em>know</em> are no longer being used.
      * 
      * <p>
-     * This method allows you to handle the special case where you are
-     * continually adding and removing the same object to and from the scene
-     * graph. In these cases, the dereference thread may see that there are no
-     * native references to the object (when it it is not in the scene graph)
-     * and remove the registered reference that lets GVRF return the original,
-     * possibly sub-classed instance. The symptom of this is code like
-     * {@link GVREyePointeeHolder#getOwnerObject()} returning a base class
-     * (<i>eg.</i> {@link GVRSceneObject}) instead of the sub-classed object
-     * that you expect. In these cases, setting the the keep-wrapper flag when
-     * you create the object will preserve object-identity.
+     * This method is {@code final} to protect the {@code close()} /
+     * {@code finalize()} 'machinery' - if you need to run any code when a
+     * descendant is closed, override the ({@code protected})
+     * {@link #destructor()} method.
+     * 
+     * @since 2.0.0
      */
-    public void setKeepWrapper(boolean keep) {
-        if (keep) {
-            mFlags |= KEEP_WRAPPER_MASK; // set bit
-        } else {
-            mFlags &= ~KEEP_WRAPPER_MASK; // clear bit
+    @Override
+    public final void close() throws IOException {
+        if (mNativePointer != 0) {
+            destructor();
+            mNativePointer = 0l;
         }
     }
 
     /**
-     * Get the current state of the keep-wrapper flag.
+     * Calls the native destructor, {@code virtual ~HybridObject()}.
+     * 
+     * This will be called only once, no matter how many times an app calls
+     * {@link #close()}. Most classes will not need to override this; the
+     * exception is classes that maintain any sort of 'registry' of 'live'
+     * classes (like {@link GVREyePointeeHolder}). If you <em>do</em> override
+     * this method, be sure to call {@code super.destructor()}.
      * 
-     * See {@link #setKeepWrapper(boolean)} for details.
+     * @since 2.0.0
      */
-    public boolean getKeepWrapper() {
-        return (mFlags & KEEP_WRAPPER_MASK) != 0;
+    protected void destructor() {
```

Comment:
Deriving classes can neglect to call super from overriding implementations, which seems like it would punch a hole through the protection that is supposed to be provided by making finalize() and close() final.